crates/biome_graphql_semantic/src/semantic_model/builder.rs

```diff
@@ -0,0 +1,151 @@
+            Declaration { range } => {
+                let binding_id = self.bindings.len();
+                self.bindings.push(SemanticModelBinding {
+                    index: binding_id.into(),
+                    range,
+                });
+                self.bindings_by_start.insert(range.start(), binding_id);
+                self.bindings_to_references.push(Vec::new());
+            }
+            Reference {
+                range,
+                declared_at: declaration_at,
+            } => {
+                let binding_id = self.bindings_by_start[&declaration_at.start()];
+                let reference_id =
+                    if let Entry::Vacant(e) = self.references_by_start.entry(range.start()) {
+                        let reference_id = self.references.len();
+
+                        self.references.push(SemanticModelReference {
+                            index: reference_id.into(),
+                            range,
+                        });
+
+                        e.insert(reference_id);
+                        self.references_to_bindings.push(Vec::new());
+                        reference_id
+                    } else {
+                        self.references_by_start[&range.start()]
+                    };
+                self.references_to_bindings[reference_id].push(binding_id);
+                self.bindings_to_references[binding_id].push(reference_id);
+            }
+
+            UnresolvedReference { range } => {
+                let node = &self.node_by_range[&range];
+                let name = node.text_trimmed().to_string();
+                if !Self::is_builtin_type(&name) && !Self::is_builtin_directive(&name) {
+                    self.unresolved_references
+                        .push(SemanticModelUnresolvedReference { range })
+                }
+            }
+            UnresolvedVariableReference {
+                range,
+                referenced_operation,
+            } => {
+                self.unresolved_variable_references
+                    .push(SemanticModelUnresolvedVariableReference {
+                        range,
+                        referenced_operation,
+                    })
+            }
+        }
+    }
+
+    #[inline]
+    pub fn build(self) -> SemanticModel {
+        let data = SemanticModelData {
+            root: self.root,
+            node_by_range: self.node_by_range,
+            bindings: self.bindings,
+            bindings_by_start: self.bindings_by_start,
+            bindings_to_references: self.bindings_to_references,
+            references: self.references,
+            references_by_start: self.references_by_start,
+            references_to_bindings: self.references_to_bindings,
+            unresolved_references: self.unresolved_references,
+            unresolved_variable_references: self.unresolved_variable_references,
+        };
+        SemanticModel::new(data)
+    }
+
+    fn is_builtin_type(name: &str) -> bool {
```

Comment:
```suggestion
    const fn is_builtin_type(name: &str) -> bool {
```

Reply:
I think matching str in const fn is not yet supported by rust. https://github.com/rust-lang/rust/issues/90237

Reply:
ah I missed that, sorry